pipelines/institution_id/02-train-model-TEMPLATE.py

```diff
@@ -227,114 +222,179 @@
 
 # COMMAND ----------
 
-df_pred = df.assign(
-    **{
-        cfg.pred_col: model.predict(df_features),
-        cfg.pred_prob_col: model.predict_proba(df_features)[
-            :, 1
-        ],  # NOTE: assumes pos_label=True ?
-    }
+# Set the number of top runs to evaluate
+top_x = 3
+optimization_metric = training_params["optimization_metric"]
+
+# Fetch all runs
+runs = mlflow.search_runs(
+    experiment_ids=[experiment_id],
+    order_by=["metrics.m DESC"],
+)
+
+# Retrieve validation metric for sorting
+search_metric = (
+    f"metrics.val_{optimization_metric}"
+    if optimization_metric in ["log_loss", "roc_auc"]
+    else f"metrics.val_{optimization_metric}_score"
+)
+
+# Sort and select top run IDs based on min/max with loss vs. score
+ascending_order = optimization_metric == "log_loss"
+top_run_ids = (
+    runs.sort_values(by=search_metric, ascending=ascending_order)
+    .iloc[:top_x]["run_id"]
+    .tolist()
```

Comment:
actually, i have a general question about the changes in the training nb: why are you running this over the top N scoring models, rather than just the best? it's hard to see what's changed in the git diff, so just... broadly unsure about what's happening and why here.

Reply:
ugh, nm, you mostly addressed in your PR message. the current eval code/loop feels like it needs to be cleaned up, but that can certainly be down outside of this PR.